### DISCUSSION TITLE
Converting a byte or hex value from a single character sliced from a numeric string

### DISCUSSION THREAD

**Discussion by Comalco** [micropython/micropython#15717] [Using MicroPython] [answered]
Folks,
I am trying to splice a string and use a single digit from it as an integer (or it can be hex) value in a I2C command.
I have a 12 digit numeric string (i.e."754621300000") that is read from a database
I want to slice this string and use each digit as a pin assignment number in a command sent to a PCF8574 to drive a relay in an array of relays.
```py
string = "745261300000"
for a in range(0, len(string)+1):          # these two lines slices each digit and prints them one by one OK
    print(string[a:a+1])

a=0                                                       # sliced digit position pointer
for y in range(32, 40):                        # Select Relay Matrix board (PCF addresses are 32 to 39)
    pcf = pcf8574.PCF8574(i2c, y)      # set this address to use
    b=(string[a:a+1])                             # slice one character at a time from the string
    print(b)                                             # print it 
    pcf.pin(b, 1)                                     # set this bit of this PCF board   (this line isnt working as I hoped)
         #  pcf.pin(7, 1)                                     # this works OK
    a=a+1                                                # increment character pointer
```
pcf.pin(b, 1)            is the problem line.  The format of b is not valid.

I saw this on a thread and thought that this would convert b into a hex value.......but it isnt seeing b as a single character string
        f="{0:02x}".format(ord(b))

I think I need to convert the single sliced character from the string to a byte in an array.  I have tried various ways to do this, but none work to provide a valid 'b' for the I2C   pcf.pin(b, 1)   command syntax.  What is the correct way to do this conversion?
As a relative newbie to MP I would appreciate some pointers or hints please? ................. TIA

**Comment by steveh127** [comment, marked as answer]
You are rather overcomplicating things. I think you will find ```int(b)``` will fix your problem.

**Reply by jomasnash** [reply]
A string is already iteratable so you can use:

```python
string = "745261300000"

for c in string:
    print(int(c))
```
in the first part of your example.

**Reply by Comalco** [reply]
Thanks, but printing isn't the issue......   its getting a usable value of the sliced character from the string so the I2C command is happy that was the issue..........

**Reply by scruss** [reply]
`int(c)` gets you the usable value of the sliced character directly

```python3
string = "745261300000"

for c in string:
    i = int(c)
    print(c, type(c), "ASCII:", ord(c), "Value:", i, type(i))
```

Output:

```
7 <class 'str'> ASCII: 55 Value: 7 <class 'int'>
4 <class 'str'> ASCII: 52 Value: 4 <class 'int'>
5 <class 'str'> ASCII: 53 Value: 5 <class 'int'>
2 <class 'str'> ASCII: 50 Value: 2 <class 'int'>
6 <class 'str'> ASCII: 54 Value: 6 <class 'int'>
1 <class 'str'> ASCII: 49 Value: 1 <class 'int'>
3 <class 'str'> ASCII: 51 Value: 3 <class 'int'>
0 <class 'str'> ASCII: 48 Value: 0 <class 'int'>
0 <class 'str'> ASCII: 48 Value: 0 <class 'int'>
0 <class 'str'> ASCII: 48 Value: 0 <class 'int'>
0 <class 'str'> ASCII: 48 Value: 0 <class 'int'>
0 <class 'str'> ASCII: 48 Value: 0 <class 'int'>
```

**Comment by sosi-deadeye** [comment]
I would make a class. To save lines, inheritance could be used or to make it easier to understand, use composition.

```python
from machine import I2C

from pcf8574 import PCF8574


# with composition
class PcfSender:
    def __init__(self, i2c, address=0x20):
        self.pcf = PCF8574(i2c, address)

    def set_pins(digits: str):
        pin_numbers = tuple(map(int, digits))

        if any(not pin < 8 for pin in pin_numbers):
             raise RuntimeError("Pin number too high")

        for pin in pin_numbers:
             self.pcf.pin(pin, 1)


i2c = I2C(0)
# 0x20 seems to be the default address of PCF if
# A0-2 is not set
pcf_sender = PcfSender(i2c, 0x20)


# with inheritance
class PcfSender(PCF8574):
    def set_pins(digits: str):
        pin_numbers = tuple(map(int, digits))

        if any(not pin < 8 for pin in pin_numbers):
             raise RuntimeError("Pin number too high")

        for pin in pin_numbers:
             self.pcf.pin(pin, 1)


i2c = I2C(0)
pcf_sender = PcfSender(i2c, 0x20)
```

This code applies for each single character the `int` function, which return an `int`, if the input is a valid digit:
```python
pin_numbers = tuple(map(int, digits))
```

**Reply by Comalco** [reply]
I will try this out later this evening, thanks.

**Comment by sosi-deadeye** [comment]
Btw. there is a flaw. For each PCF you only have 8 ports. This means if you want to set the Pins of one PCF, then you could only have a maximum of 8 digits. But your example holds 12 digits. Do you have more than one PCF and want to set them?

**Reply by Comalco** [reply]
The ''for y in range(32, 40)'' routine steps through 8 possible PCF addresses.  One value from the string is written to each PCF.   And I have more than 8 PCFs.  There are actually 24 PCFs, driving a matrix of 144 relays, of which one relay per 2 PCFs is written with a value from the string.  A zero in the string means no relay is enabled for that PCF.  Other pins of the many PCFs drive LEDs etc, which are separate to the relay configuration string which is retrieved from a database..  All of my I2C code, addressing etc works OK.  It is slicing the string into single characters and getting that into a value format for the I2C command that is the issue.  Many thanks.

**Comment by Comalco** [comment]
This is working for me.  Perhaps not optimal, but working.  I will clean it up as I learn more.  Thanks.
```py
string = "745261380000"
a=0
for y in range(32, 40, 2):             # Select Relay Matrix board (PCF address)
        b1=(string[a:a+1])
        l = ord(b1)
        b = bytearray(l)
        b[0]=ord(b1)
        if b[0]-48 > 7:
            pcf = pcf8574.PCF8574(i2c, y)
            pcf.pin(b[0]-49, 1)
        elif b[0]-48 > 0:    
            pcf = pcf8574.PCF8574(i2c, y)
            pcf.pin(b[0]-49, 1)
        elif b[0]-48 <= 0:    
            y += 2
            pcf = pcf8574.PCF8574(i2c, y)
            pcf.pin(b[0]-42, 1)
        if w == 1 and y >= 35:
            break
        a+=1
'''

**Comment by Comalco** [comment]
I finished up needing to utilise some other digits as well as 0-9 in my string.
I did an iteration using ord() and a few other changes along the way.   In the end I settled on some working code that catered for everything I needed.  Many thanks to all for helpful inputs.
